### ISSUE TITLE
ngOnDestroy not fired when service is provided using useFactory

### ISSUE THREAD

**Issue by colinkahn** [angular/angular#14818]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When you provide a service using `useFactory`, `ngOnDestroy` will never be called.

**Expected behavior**
Regardless of how you provide the service, `ngOnDestroy` should get called if it is a method on the service.

**Minimal reproduction of the problem with instructions**
https://plnkr.co/edit/aUaz6ewyM3m5JorsAoCO?p=preview

1. Open the console.
2. Click the "Create Component" button one or many times
3. After a 2 second delay the component will be destroyed, but no console output will be seen.
4. If you replace `useFactory` with `useClass` the problem goes away and you'll see output.

**What is the motivation / use case for changing the behavior?**
There are many cases, main motivator is that this isn't intuitive.

**Please tell us about your environment:**

* **Angular version:** 2.4.8

* **Browser:** Chrome 56.0.2924.87

 
* **Language:** TypeScript 2.0

* **Node (for AoT issues):** `node --version` =

**Comment by DzmitryShylovich** [comment]
it's by design. there's no way to know if returned by the factory object has `ngOnDestroy` or not.

**Comment by colinkahn** [comment]
As a user `useClass`, `useValue`, and `useFactory` would seem to all just be different ways to achieve the same result. I think this should be a feature if it's not possible in the current implementation.

**Comment by colinkahn** [comment]
A few other related examples showing that even `useClass` doesn't always work:

https://plnkr.co/edit/eIs7pjvldhcFrMmaJ4IV?p=preview

In that version I'm providing an abstract class and using `useClass` to specify a concrete version. `ngOnDestroy` is never called.

https://plnkr.co/edit/G7vLIvA1nz1dBoJ4sPcW?p=preview

Similar to the above, but using an OpaqueToken.

As far as I can tell the only time `ngOnDestroy` does work is if you use one of these (I believe they're equivalent):

```
[
  {
    provide: X,
    useClass: X,
  },
  // OR
  X,
]
```

**Comment by DzmitryShylovich** [comment]
> In that version I'm providing an abstract class and using useClass to specify a concrete version. ngOnDestroy is never called.

this is looks like a valid bug

**Comment by DzmitryShylovich** [comment]
can you mark this as a feature request and file a new issue for https://github.com/angular/angular/issues/14818#issuecomment-283200005 ?

**Comment by SonyStone** [comment]
I have simuler issuer with Factory providers from this gide:

https://angular.io/guide/dependency-injection#factory-providers

__Minimal reproduction of the problem with instructions__
https://stackblitz.com/edit/angular-service-factory-issue?file=src%2Fapp%2Ftest.service.ts

* Click the "toggle" button one or many times
   `Test 1 Service` will be created and destroyed. `Test 2 Service` will be created and never destroyed.

**Comment by simeyla** [comment]
I can't believe it doesn't work with an `InjectionToken` (used to be `OpaqueToken`).
That was very unexpected!

I need to unsubscribe to some observables in my service - I ended up having to inject a 'SubscriptionManager' service to my component and had the dependent service use that.

I suppose in future maybe it's better to have the component call `init` on the service (to return Subscription instance) and then track the subscriptions in the component.

**Comment by lloydaf** [comment]
@simeyla do you have any gists to demonstrate using a subscription manager?

**Comment by simeyla** [comment]
@lloydaf I'm not particularly happy with the way mine works - the boiler plate to just use it is equal to the effort to just manually create a Subscription.

The goal was to just make an Injectable that a component would provide to itself and you would 'add' subscriptions to it. First problem is that you can't disable inheritance for DI objects - so if you forget to provide it then you'll get the parent manager. So I added a mechanism to make sure it was initialized only once - but then you have to call init() in the constructor which is just ugly. It also put all the subscriptions in a global manager which in the end didn't really bring me the other benefits I was hoping for (such as counting subscriptions and reporting if you weren't closing them properly).

So tbh my 'subscription manager' ended up kind of a mess and I'm phasing it out where I did use it. A simple 'Subscription' object on the class or using something like [takeUntilDestroy](https://github.com/NetanelBasal/angular2-take-until-destroy) works better. Also there are more issues with it that are even more complicated and 'providing' it to you would be more of a foot gun than anything else.

If you're only using ngOnDestroy to unsubscribe to observables you can do one of two things:

1) Only allow the component to own the Subscription - eg. coding standard says services can't subscribe.

2) I just came up with this and I actually kind of like it for my situation. Simple and clear and relatively safe.

`export const COMPONENT_SUBSCRIPTION = new InjectionToken<Subscription>("COMPONENT_SUBSCRIPTION"); `

Then a factory function to create it:

`export const componentSubscriptionFactory = () => new Subscription();`

In every component you must provide it - and be sure to use @Self to prevent inheriting a parent subscription (big mess if you don't):

```
        {
            provide: COMPONENT_SUBSCRIPTION,
            useFactory: componentSubscriptionFactory
        }
```

`@Self() @Inject(COMPONENT_SUBSCRIPTION) private subscription: Subscription,`

You must also put this in the component (only)

```
ngOnDestroy() 
{
    debugger;  // just to prove it works!
    this.subscription.unsubscribe();
}
```

and in services you use this:

`@Inject(COMPONENT_SUBSCRIPTION) subscription: Subscription`

The service will then get your component's `Subscription` object and it's just a normal RxJS object you can call `add(...)` on.

And none of this is any help if you need to dispose of anything other than RxJS subscriptions!
